File: Problems/Kattis/Zagrade.cpp
#include <bits/stdc++.h>
using namespace std;

#define rep(i, n) for (int i = 0; i < (int)n; i++)

int N;
string S;
vector<pair<int, int>> P;

int main()
{
    cin >> S; N = S.size();

    stack<int> s;
    rep(i, N)
    {
        if (S[i] == '(') s.push(i);
        if (S[i] == ')') { P.emplace_back(s.top(), i); s.pop(); }
    }

    int M = P.size();

    set<string> ans;
    rep(b, (1 << M) - 1)
    {
        set<int> out;
        rep(i, M) if (!(b & (1 << i))) out.insert(P[i].first), out.insert(P[i].second);

        string aux;
        rep(i, N) if (!out.count(i)) aux += S[i];

        ans.insert(aux);
    }

    for (string aux : ans) cout << aux << '\n';
}

File: Problems/Kattis/Zagrade.py

S = input().strip()
N = len(S)

s = []
P = []
for i in range(N):
    if S[i] == '(':
        s.append(i)
    if S[i] == ')':
        P.append((s.pop(), i))

M = len(P)

ans = set()
for b in range((1 << M) - 1):
    out = set()
    for i in range(M):
        if not (b & (1 << i)):
            out.add(P[i][0])
            out.add(P[i][1])
    
    aux = ""
    for i in range(N):
        if i not in out:
            aux += S[i]

    ans.add(aux)

for aux in sorted(list(ans)):
    print(aux)
